barretenberg/cpp/src/barretenberg/ultra_honk/ultra_prover.cpp

```diff
@@ -33,65 +34,31 @@ UltraProver_<Flavor>::UltraProver_(Builder& circuit)
- *
- */
-template <IsUltraFlavor Flavor> void UltraProver_<Flavor>::execute_relation_check_rounds()
+template <IsUltraFlavor Flavor> HonkProof UltraProver_<Flavor>::export_proof()
 {
-    using Sumcheck = SumcheckProver<Flavor>;
-    auto circuit_size = instance->proving_key->circuit_size;
-    auto sumcheck = Sumcheck(circuit_size, transcript);
-    RelationSeparator alphas;
-    for (size_t idx = 0; idx < alphas.size(); idx++) {
-        alphas[idx] = transcript->template get_challenge<FF>("Sumcheck:alpha_" + std::to_string(idx));
-    }
-    instance->alphas = alphas;
-    std::vector<FF> gate_challenges(numeric::get_msb(circuit_size));
+    proof = transcript->proof_data;
+    return proof;
+}
+template <IsUltraFlavor Flavor> void UltraProver_<Flavor>::generate_gate_challenges()
+{
+    std::vector<FF> gate_challenges(numeric::get_msb(instance->proving_key->circuit_size));
     for (size_t idx = 0; idx < gate_challenges.size(); idx++) {
         gate_challenges[idx] = transcript->template get_challenge<FF>("Sumcheck:gate_challenge_" + std::to_string(idx));
     }
     instance->gate_challenges = gate_challenges;
-    sumcheck_output = sumcheck.prove(instance);
-}
-
-/**
- * @brief Execute the ZeroMorph protocol to prove the multilinear evaluations produced by Sumcheck
- * @details See https://hackmd.io/dlf9xEwhTQyE3hiGbq4FsA?view for a complete description of the unrolled protocol.
- *
- * */
-template <IsUltraFlavor Flavor> void UltraProver_<Flavor>::execute_zeromorph_rounds()
-{
-    ZeroMorph::prove(instance->prover_polynomials.get_unshifted(),
-                     instance->prover_polynomials.get_to_be_shifted(),
-                     sumcheck_output.claimed_evaluations.get_unshifted(),
-                     sumcheck_output.claimed_evaluations.get_shifted(),
-                     sumcheck_output.challenge,
-                     commitment_key,
-                     transcript);
 }
 
-template <IsUltraFlavor Flavor> HonkProof& UltraProver_<Flavor>::export_proof()
+template <IsUltraFlavor Flavor> HonkProof UltraProver_<Flavor>::construct_proof()
 {
-    proof = transcript->proof_data;
-    return proof;
-}
-
-template <IsUltraFlavor Flavor> HonkProof& UltraProver_<Flavor>::construct_proof()
-{
-    auto [relation_params] = oink_prover.prove();
+    auto [relation_params, alphas] = oink_prover.prove();
     instance->relation_parameters = std::move(relation_params);
+    instance->alphas = alphas;
     instance->prover_polynomials = ProverPolynomials(instance->proving_key);
 
-    // Fiat-Shamir: alpha
-    // Run sumcheck subprotocol.
-    execute_relation_check_rounds();
-
-    // Fiat-Shamir: rho, y, x, z
-    // Execute Zeromorph multilinear PCS
-    execute_zeromorph_rounds();
+    generate_gate_challenges();
 
-    return export_proof();
+    DeciderProver_<Flavor> decider_prover(instance, transcript);
+    return decider_prover.construct_proof();
```

Comment:
I think it's quite odd that the decider_prover is being called here since the ultra prover should not care about folding, this is one of those place where round abstraction would be awesome :-?

Reply:
but until then I would kinda leave the sumcheck and zeromorph rounds unshared between the decider and ultra prover just to not create this dependence